res/controllers/midi-components-0.1.js

```diff
@@ -0,0 +1,1135 @@
+                        if (Math.abs(change) >= this.changeThreshold
+                            // this.valueAtLastEffectSwitch can be undefined if
+                            // shift was pressed before the first MIDI value was received.
+                            || this.valueAtLastEffectSwitch === undefined) {
+                            const effectGroup = "[EffectRack1_EffectUnit" +
+                                            eu.currentUnitNumber + "_Effect" +
+                                            this.number + "]";
+                            engine.setValue(effectGroup, "effect_selector", change);
+                            this.valueAtLastEffectSwitch = value;
+                        }
+
+                        this.previousValueReceived = value;
+                    };
+                }
+                connect() {
+                    this.connections[0] = engine.makeConnection(eu.group, "focused_effect",
+                        this.onFocusChange.bind(this));
+                }
+                disconnect() {
+                    engine.softTakeoverIgnoreNextValue(this.group, this.inKey);
+                    if (this.connections[0] !== undefined) {
+                        this.connections[0].disconnect();
+                    }
+                }
+                trigger() {
+                    this.connections[0].trigger();
+                }
+                onFocusChange(value, _group, _control) {
+                    if (value === 0) {
+                        this.group = "[EffectRack1_EffectUnit" +
+                                    eu.currentUnitNumber + "_Effect" +
+                                    this.number + "]";
+                        this.inKey = "meta";
+                    } else {
+                        this.group = "[EffectRack1_EffectUnit" + eu.currentUnitNumber +
+                                    "_Effect" + value + "]";
+                        this.inKey = "parameter" + this.number;
+                    }
+                    engine.softTakeoverIgnoreNextValue(this.group, this.inKey);
+                }
+            };
+
+            this.EffectEnableButton = class extends Button {
+                constructor(number) {
+                    const options = {
+                        number: number,
+                        group: "[EffectRack1_EffectUnit" + eu.currentUnitNumber +
+                                "_Effect" + number + "]",
+                        type: Button.types.powerWindow,
+                    };
+                    super(options);
+                }
+                // NOTE: This function is only connected when not in focus choosing mode.
+                onFocusChange(value, _group, _control) {
+                    if (value === 0) {
+                        if (colors !== undefined) {
+                            this.color = colors.unfocused;
+                        }
+                        this.group = "[EffectRack1_EffectUnit" +
```

Comment:
we see this building of the group name over and over again -- can we factor that out into a function effectGroupName() or some sort?

Reply:
I think this is the last bit of low-hanging fruit we can do now.  Does that seem reasonable?